File: src/main/java/com/senderman/stickersorterbot/bot/command/CommandExtractor.java
package com.senderman.stickersorterbot.bot.command;


import javax.annotation.Nullable;

public interface CommandExtractor {

    /**
     * Find executor by command
     *
     * @param command command starting with "/"
     * @return appropriate command executor implementation, or null if command not found
     */
    @Nullable
    public CommandExecutor findExecutor(String command);

}

File: src/main/kotlin/com/senderman/stickersorterbot/bot/CommandExtractor.kt
package com.senderman.stickersorterbot.bot

import org.springframework.beans.factory.NoSuchBeanDefinitionException
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Component

/**
 * Class that can be used to get CommandExecutor by command.
 * To make things work, annotate each CommandExecutor interface implementation as @Component(command)
 * For example: @Component("/help") class Help{...}
 */
@Component
class CommandExtractor(
        private val context: ApplicationContext
) {

    fun findExecutor(command: String): CommandExecutor? =
            try {
                context.getBean(command, CommandExecutor::class.java)
            } catch (e: NoSuchBeanDefinitionException) {
                null
            }
}